src/library/scala/util/control/TailCalls.scala

```diff
@@ -111,4 +111,31 @@ object TailCalls {
    */
   def done[A](result: A): TailRec[A] = Done(result)
 
+
+  /**
+   * Call a function with a state argument, and a function which will
+   * recursively call the passed function with the new state.
+   *
+   * This can be used to implement a loop similar to using self-recursion
+   * with an anonymous function.
+   *
+   * Fore example computing a fibonacci number could be implemented as:
+   *
+   * {{{
+   * import scala.util.control.TailCalls._
+   *
+   * def fib(n: Int): Int =
+   *    recurse[(Int, Int, Int), Int]((n, 1, 0)) ( recur => { case (i, n1, n2) =>
```

Comment:
It's unfortunate that you have to explicitly give the type parameters, but I wasn't able to figure out a way to get scala to infer the types (at least in scala 2, maybe 3 can do it?)

Needing an extra set of braces around the case is a little annoying too.

Reply:
You can make a Recursor class that captures the initial state and infers the type, then that class has an apply method accepts the lambda